### ISSUE TITLE
Newlines without Unicode

### ISSUE THREAD

**Issue by dbohdan** [jointwt/twtxt#215]
## Is your feature request related to a problem? Please describe.
twtxt.net has implemented newlines (#166).  It is a desirable feature: users of competing products write many statuses with newlines.  However, the current implementation depends on clients being able to read and write UTF-8 and understanding the convention of rendering [\u2028](https://www.fileformat.info/info/unicode/char/2028/index.htm) as a newline.  Software that supports UTF-8 but is unaware of the convention seems to render \u2028 as:
* nothing (Firefox, xterm);
* an arrow (Guake);
* `<U+2028>` ([bat](https://github.com/sharkdp/bat));
* [mojibake](https://en.wikipedia.org/wiki/Mojibake) or a truncated message ([buckket/twtxt](https://github.com/buckket/twtxt)).

This means twtxt.net feeds following the convention look wrong in other clients.  UTF-8 is also difficult to work with on retro computers, a niche where twtxt may be adopted.

## Describe the solution you'd like
One can implement newlines in twtxt in a backwards-compatible manner that does not require Unicode input: **when the client sees a sequence of twts with the same timestamp, have it join their text with a newline**.  Feeds following this convention look if not quite correct then reasonable in clients that do not understand it (as long as they display twts with the same timestamp in the order they appear).  See the example below.  Creating twts that follow the same-timestamp convention is trivial.

The convention can be implemented alongside with or instead of #166.

## Describe alternatives you've considered
Other non-Unicode options:
* Translating `\n` to newlines.  This would break code embedded in twts.  I don't think it's a good idea.
* Letting the user escape a newline with a `\` at the end of a line.  This breaks backwards compatibility.

## Example

buckket/twtxt v1.2.3 renders

```none
1845-01-29T12:00:00Z	Once upon a midnight dreary, while I pondered, weak and weary,
1845-01-29T12:00:00Z	Over many a quaint and curious volume of forgotten lore—
1845-01-29T12:00:00Z	    While I nodded, nearly napping, suddenly there came a tapping,
1845-01-29T12:00:00Z	As of some one gently rapping, rapping at my chamber door.
1845-01-29T12:00:00Z	“’Tis some visitor,” I muttered, “tapping at my chamber door—
1845-01-29T12:00:00Z	            Only this and nothing more.”
```

as

```none
➤ http://127.0.0.1:8081/poe.txt (175 years ago):
Once upon a midnight dreary, while I pondered, weak and weary,

➤ http://127.0.0.1:8081/poe.txt (175 years ago):
Over many a quaint and curious volume of forgotten lore—

➤ http://127.0.0.1:8081/poe.txt (175 years ago):
While I nodded, nearly napping, suddenly there came a tapping,

➤ http://127.0.0.1:8081/poe.txt (175 years ago):
As of some one gently rapping, rapping at my chamber door.

➤ http://127.0.0.1:8081/poe.txt (175 years ago):
“’Tis some visitor,” I muttered, “tapping at my chamber door—

➤ http://127.0.0.1:8081/poe.txt (175 years ago):
Only this and nothing more.”
```

**Comment by dbohdan** [comment]
I've just implemented this in the [twtxt.tcl](https://gitlab.com/dbohdan/twtxt.tcl) library and reader.

![Timeline reader screenshot](https://user-images.githubusercontent.com/3179832/93906150-8453c380-fd04-11ea-9210-debfb964eb7b.png)

**Comment by prologic** [comment]
This is cool! Are you willing to put up a PR here that changes this behaviour? Be very keen to see it working here...

**Comment by dbohdan** [comment]
Thanks!  I think somebody who actually works with the codebase should implement it to avoid breaking the parser*.  Realistically, this means you, @prologic. :sweat_smile: 

\* Which could use unit tests!  I like the philosophy of "Write tests. Not too many. Mostly integration."  However, when it comes to parsers unit tests help a lot.

**Comment by dbohdan** [comment]
I've made a [standalone page for my proposal](https://dbohdan.com/wiki?name=Newlines%20in%20twtxt).

**Comment by prologic** [comment]
The only part I find weird about this proposal is blank lines or new paragraphs if you will, e.g:

```
1845-01-29T12:00:00Z	Once upon a midnight dreary, while I pondered, weak and weary,
1845-01-29T12:00:00Z	Over many a quaint and curious volume of forgotten lore—
1845-01-29T12:00:00Z	    While I nodded, nearly napping, suddenly there came a tapping,
1845-01-29T12:00:00Z	
1845-01-29T12:00:00Z	As of some one gently rapping, rapping at my chamber door.
1845-01-29T12:00:00Z	“’Tis some visitor,” I muttered, “tapping at my chamber door—
1845-01-29T12:00:00Z	
1845-01-29T12:00:00Z	            Only this and nothing more.”
```

Of course this will still work and render corr3ectly, no problems there.

No big deal?

**Comment by dbohdan** [comment]
Users probably want to insert blank lines in their multiline twts.  I think it looks fine when you display them:

![twtxt.tcl timeline reader screenshot](https://user-images.githubusercontent.com/3179832/95058031-a0d9fd80-06ff-11eb-842e-254894f09749.png)

As for single-line blank twts, the format spec says nothing about them.  The original twtxt client does not display blank twts.  I display them in my feed reader because someone may use them as a form of notification or something like that.  My general policy with such things is that if it is there in the file and parses as a twt, it should be displayed.

Something a client may choose to do when rendering a twt is prevent extremely tall multiline posts.  Those can ruin a user's view of a timeline.  I see several options for this:

1. You can collapse vertical whitespace: treat every `\n+` as a single `\n`.  This is what John Gruber's original Markdown.pl and the C reference implementation of CommonMark in its default mode do.
2. You can limit the number of lines in a twt and either truncate the twt or give it a scrollbar or the equivalent of a scrollbar when the twt exceeds _N_ lines.  (10? 25?)
3. You can forcibly stop joining twts with the same timestamp after _N_ lines, breaking up a long twt into several.  I would not choose this option.
4. You can combine 1 and 2.

**Comment by prologic** [comment]
@dbohdan The only reason I haven't implemented this yet is it actually complicates the parser a bit... I have to do funky stuff like keeping a map of timestamps or a buffer so I can combine multiple lines together...

**Comment by dbohdan** [comment]
I don't think you need to (or should) merge twts with the same timestamp when they do not occur in sequence.  With this constraint you don't need a map or a buffer.  See [how I do it](https://gitlab.com/dbohdan/twtxt.tcl/-/blob/d609d524311678e7b81866769a91237e4bc19e47/lib/twtxt/twtxt.tcl#L91) in my parser.  The code is in Tcl, but the general logic should be understandable without knowing the language.  We store the timestamp of the previous twt.  When we parse a line, we check if the timestamp is the same as the previous.  If it is, we take the last twt in our already-parsed list and append a newline and the text of the current twt to its text.

**Comment by prologic** [comment]
> We store the timestamp of the previous twt. When we parse a line, we check if the timestamp is the same as the previous. If it is, we take the last twt in our already-parsed list and append a newline and the text of the current twt to its text.

Ahh I feel dumb now 🤣 I didn't think this through properly, you are right! This should be easy to implement! 🙇‍♂️

**Comment by prologic** [comment]
@dbohdan Sorry for the delay on this issue, I just _realized_ today (_not sure why it took me this long to realize?! 😳_) that by breaking multi-line posts over multiple Twts (_lines_), we effectively break a rather important feature of an "extension" we built (_blogged about it [here](https://twtxt.net/blog/prologic/2020/10/18/making-twtxt-better)_).

That is Hash(ability) of the Twt itself would now be broken if we combined multiple lines with the same timestamp into a single Twt.

Although now that I write this (_using Github Issues as a "rubber duck" 🐤_), it _may_ be "okay" perhaps as long as other clients also adopt the Hash/Reply conventions as well as the Multiline conventions.

I _think_ its time to document:

- Twt Hashes
- Twt Replies and Conversations
- Multiline Twts

At [dev.twtxt.net](https://dev.twtxt.net)

What do you think about all this?

**Comment by dbohdan** [comment]
I think that for addressing and hashing we should consider twts data structures in the program's memory (regardless of how they got there), not lines in a twtxt file.  That is to say, if you implement merging, merged twts should be addressed and hashed as one and their constituent twts should not be individually hashed or addressable.  Addressing should happen after and be separate from parsing.  You should be able to hook up an additional parser (e.g., if a 50-page YAML-based twtxt v2.0 standard comes out) and use it right away with the other features.

**Comment by prologic** [comment]
Yeah as I said, as I was writing my comment I came to a similar conclusion as you. So I _think_ its fine.

The addressing is an important aspect of how we've extended the "usefulness" a bit, and I _believe_ you came up with it in fact? 🙇‍♂️

**Comment by dbohdan** [comment]
Yes, I PRed it in #44.  I see that you have since shortened the ids, which is a good idea.
